src/components/views/rooms/RoomHeader/CallGuestLinkButton.tsx

```diff
@@ -0,0 +1,170 @@
+import ShareDialog from "../../dialogs/ShareDialog";
+import { _t } from "../../../../languageHandler";
+import SettingsStore from "../../../../settings/SettingsStore";
+import { calculateRoomVia } from "../../../../utils/permalinks/Permalinks";
+import BaseDialog from "../../dialogs/BaseDialog";
+import { useGuestAccessInformation } from "../../../../hooks/room/useGuestAccessInformation";
+
+/**
+ * Display a button to open a dialog to share a link to the call using a element call guest spa url (`element_call:guest_spa_url` in the EW config).
+ * @param room
+ * @returns Nothing if there is not the option to share a link (No guest_spa_url is set) or a button to open a dialog to share the link.
+ */
+export const CallGuestLinkButton: React.FC<{ room: Room }> = ({ room }) => {
+    const { canInviteGuests, guestSpaUrl, isRoomJoinable, canInvite, isRoomJoinableFunction } =
+        useGuestAccessInformation(room);
+
+    const generateCallLink = useCallback(() => {
+        if (!isRoomJoinableFunction())
+            throw new Error("Cannot create link for room that users can not join without invite.");
+        if (!guestSpaUrl) throw new Error("No guest SPA url for external links provided.");
+        const url = new URL(guestSpaUrl);
+        url.pathname = "/room/";
+        // Set params for the sharable url
+        url.searchParams.set("roomId", room.roomId);
+        if (room.hasEncryptionStateEvent()) url.searchParams.set("perParticipantE2EE", "true");
+        for (const server of calculateRoomVia(room)) {
+            url.searchParams.set("viaServers", server);
+        }
+
+        // Move params into hash
+        url.hash = "/" + room.name + url.search;
+        url.search = "";
+
+        logger.info("Generated element call external url:", url);
+        return url;
+    }, [guestSpaUrl, isRoomJoinableFunction, room]);
+
+    const showLinkModal = useCallback(() => {
+        try {
+            // generateCallLink throws if the invite rules are not met
+            const target = generateCallLink();
+            Modal.createDialog(ShareDialog, {
+                target,
+                customTitle: _t("share|share_call"),
+                subtitle: _t("share|share_call_subtitle"),
+            });
+        } catch (e) {
+            logger.error("Could not generate call link.", e);
+        }
+    }, [generateCallLink]);
+
+    const shareClick = useCallback(() => {
+        if (isRoomJoinable) {
+            showLinkModal();
+        } else {
+            // the room needs to be set to public or knock to generate a link
+            Modal.createDialog(JoinRuleDialog, {
+                room,
+                // If the user cannot invite the Knocking is not given as an option.
+                canInvite,
+            }).finished.then(() => {
+                // we need to use the function here because the callback got called before the state was updated.
+                if (isRoomJoinableFunction()) showLinkModal();
```

Comment:
Is there some reason I'm not seeing why we need `isRoomJoinable` in addition to `isRoomJoinableFunction`? If the `useGuestAccessInformation` hook only exposed the latter, I think that would work as well.

Reply:
isRoomJoinableFunction is needed because it could change while we are still in the showModal callback (because we chain the modals.)

Bet we need the state isRoomJoinable for having reactivity in the buttons (the share link button can appear/disappear on the fly if the power levels change)

Also isRoomJoinableFunction needs to do the full room state check and is a bit more expensive to compute. Where isRoomJoinable is based only on state variables using useRoomState that will be updated anyways.

Reply:
I double checked and since it is only used at one location (and that is in the click callback) the reactivity of the `isRoomJoinable` field is not requiered anymore. also the performance penalty is basicially nothing since we only execute it once for each click.

I forgot that I am not relying on `isRoomJoinable` for any rendering code anymore its all handled with: 
```
canInviteGuests, guestSpaUrl, canInvite
```
now.
Which internally depends on isRoomJoinable but we really dont need to expose it.